File: FlinkScalaCode/src/main/scala/com/wubaibao/flinkscala/code/chapter9/dsandtableintegration/dstotable/FromChangelogStreamTest.scala
package com.wubaibao.flinkscala.code.chapter9.dsandtableintegration.dstotable

import org.apache.flink.api.common.typeinfo.Types
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.{Schema, Table}
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
import org.apache.flink.table.connector.ChangelogMode
import org.apache.flink.types.{Row, RowKind}


/**
 * Flink DataStream 转换为 Table
 * 使用 tableEnv.fromChangelogStream() 方法将 DataStream 转换为 Table
 */
object FromChangelogStreamTest {
  def main(args: Array[String]): Unit = {
    //创建流处理执行环境
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    //导入隐式转换
    import org.apache.flink.streaming.api.scala._

    //创建TableEnv
    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)


    val dataStream:
      DataStream[Row] = env.fromElements(
      Row.ofKind(RowKind.INSERT, "zs", Int.box(18)),
      Row.ofKind(RowKind.INSERT, "ls", Int.box(19)),
      Row.ofKind(RowKind.UPDATE_BEFORE, "zs", Int.box(18)),
      Row.ofKind(RowKind.UPDATE_AFTER, "zs", Int.box(20)),
      Row.ofKind(RowKind.DELETE, "ls", Int.box(19))
    )(Types.ROW(Types.STRING, Types.INT))

    //将DataStream 转换成 Table
    val result: Table = tableEnv.fromChangelogStream(
      dataStream,
      //通过Schema指定主键
      Schema.newBuilder.primaryKey("f0").build,
      //指定ChangelogMode，这里使用all()，表示所有类型的数据都会被处理
      ChangelogMode.all()
    )

    //打印表结构
    result.printSchema()

    //打印表数据
    result.execute.print()

  }

}

File: FlinkJavaCode/src/main/java/com/wubaibao/flinkjava/code/chapter9/dsandtableintegration/dstotable/FromChangelogStreamTest.java
package com.wubaibao.flinkjava.code.chapter9.dsandtableintegration.dstotable;

import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Schema;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.connector.ChangelogMode;
import org.apache.flink.types.Row;
import org.apache.flink.types.RowKind;

/**
 * Flink DataStream 转换为 Table
 * 使用 tableEnv.fromChangelogStream() 方法将 DataStream 转换为 Table
 */
public class FromChangelogStreamTest {
    public static void main(String[] args) {
        //创建流处理执行环境
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //为了看出数据顺序效果，设置并行度为1
        env.setParallelism(1);

        //创建TableEnv
        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

        DataStream<Row> dataStream =
                env.fromElements(
                        Row.ofKind(RowKind.INSERT, "zs", 18),
                        Row.ofKind(RowKind.INSERT, "ls", 19),
                        Row.ofKind(RowKind.INSERT, "ww", 20),
                        Row.ofKind(RowKind.UPDATE_BEFORE, "zs", 18),
                        Row.ofKind(RowKind.UPDATE_AFTER, "zs", 20),
                        Row.ofKind(RowKind.DELETE, "ls", 19),
                        Row.ofKind(RowKind.UPDATE_AFTER, "ww", 21),
                        Row.ofKind(RowKind.INSERT, "zs", 18)
                        );

        //将DataStream 转换成 Table
        Table result = tableEnv.fromChangelogStream(dataStream,
                //通过Schema指定主键
                Schema.newBuilder().primaryKey("f0").build(),
                //指定ChangelogMode，这里使用all()，表示所有类型的数据都会被处理
                ChangelogMode.all()
        );

        //打印表结构
        result.execute().print();
    }
}
